File: packages/bmsx-console/src/render/host_overlay/overlay_queue.ts
import type { Host2DSubmission } from '../shared/submissions';

export type HostOverlayFrame = {
	width: number;
	height: number;
	logicalWidth: number;
	logicalHeight: number;
	renderWidth: number;
	renderHeight: number;
	commands: Host2DSubmission[];
};

let pendingFrame: HostOverlayFrame = null;

export function publishOverlayFrame(frame: HostOverlayFrame): void {
	pendingFrame = frame;
}

export function hasPendingOverlayFrame(): boolean {
	return pendingFrame !== null;
}

export function consumeOverlayFrame(): HostOverlayFrame {
	const frame = pendingFrame;
	pendingFrame = null;
	return frame;
}

export function clearOverlayFrame(): void {
	pendingFrame = null;
}

File: native/machine/render/host_overlay/overlay_queue.cpp
#include "render/host_overlay/overlay_queue.h"

namespace bmsx {
namespace {

HostOverlayFrame g_pendingFrame;
bool g_hasPendingFrame = false;

} // namespace

void publishOverlayFrame(const HostOverlayFrame& frame) {
	g_pendingFrame = frame;
	g_hasPendingFrame = true;
}

bool hasPendingOverlayFrame() {
	return g_hasPendingFrame;
}

HostOverlayFrame consumeOverlayFrame() {
	g_hasPendingFrame = false;
	return g_pendingFrame;
}

void clearOverlayFrame() {
	g_hasPendingFrame = false;
}

} // namespace bmsx
